ctsm5.3.059: Various cleanup efforts from the cesm3_0_beta04 tags for testing and usability

ekluzek opened the pull request:
### Description of changes

Various updates for testing and other problems identified in the cesm3_0_beta04 tag. So fixes and cleanup for usability.
Including the following:

- Fix SHR_ASSERT so single-point matrix test passes
- ne3np4 to namelist_defaults_ctsm.xml and Makefile for PTS mode and add ability for warm starts in PTS mode 
- f19 + f45 16pft fsurdat/landuse files to namelist_defaults_ctsm + Makefile
- Changes in the FORTRAN code to properly abort when fire-emission is asked for it can't be provided. Added unit testing for this.

### Specific notes

Contributors other than yourself, if any: @slevis-lmwg 

CTSM Issues Fixed (include github issue #):
  Fixes #2791
  Fixes #2768
  Fixes #2780
  Fixes #2762
  Fixes #3073
  Some of #2810
CTSM namelist checking for: https://github.com/NGEET/fates/issues/1356
Some work on https://github.com/ESCOMP/CTSM/issues/2643

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? Yes

Does this create a need to change or add documentation? Did you do so? No 

Testing performed, if any: regular

PR's Involved:

#2840
#2835
#2834
#2844

ekluzek committed:
Add tests for fire_emis with SP bgc mode or BGC mode with nofire, they fail as expected

ekluzek committed:
Make CLM_BLDNML_OPTS explicitly turn off drv_flds_in options off for FATES, and also off when coupled to CAM

slevis-lmwg committed:
Update namelist_defaults with ne3np4 fsurdat/landuse files

slevis-lmwg committed:
Update namelist_defaults with ne3np4 fsurdat/landuse files

slevis-lmwg committed:
Add ne3np4 fsurdat/landuse generation to Makefile

slevis-lmwg committed:
Add ne3np4 fsurdat/landuse generation to Makefile

slevis-lmwg committed:
Small important correction to namelist_defaults_ctsm.xml

slevis-lmwg committed:
Small important correction to namelist_defaults_ctsm.xml

slevis-lmwg committed:
Add build-namelist tests for ne3np4 same as existing ne3np4.pg3

slevis-lmwg committed:
Add build-namelist tests for ne3np4 same as existing ne3np4.pg3

ekluzek committed:
Start adding a unit tests for CNFireFactory with just a single simple initialization of the CNFireNorFire class

ekluzek committed:
Start adding a unit tests for CNFireFactory with just a single simple initialization of the CNFireNorFire class

ekluzek committed:
Merge remote-tracking branch 'escomp/cesm3_0_beta04_changes' into fire_emis

slevis-lmwg committed:
Change ptsRLA and ptsRLO tests to ne3 in testlist_clm.xml

slevis-lmwg committed:
Merge remote-tracking branch 'slevis-lmwg/ne3np4_to_defaults_and_makefile' into ne3np4_to_defaults_and_makefile

slevis-lmwg committed:
Updating the ptsRLA and ptsRLO tests to the correct ne3 grid

slevis-lmwg committed:
Merge branch 'cesm3_0_beta04_changes' into ne3np4_to_defaults_and_makefile

slevis resolved conflicts:
bld/unit_testers/build-namelist_test.pl

jtruesdal committed:
updates for ne3np4 scam

slevis-lmwg committed:
Merge branch 'cesm3_0_beta04_changes' into ne3np4_to_defaults_and_makefile

jtruesdal committed:
ctsm warm_start mods

slevis-lmwg committed:
Bring back clm5-cam6 init_interp_attributes for ne0CONUS grid

slevis-lmwg committed:
Merge pull request #2901 from slevis-lmwg/fix_failing_ne0_tests

Bring back clm5-cam6 init_interp_attributes for ne0CONUS grid

jtruesdal committed:
bug fix for scam clm warmstart - SCT cam6_prep now produces BFB results with 3d CAM using F2000climo with SROF

slevis-lmwg committed:
Merge pull request #17 from jtruesdal/ctsm_warmstart

ctsm warm_start mods

slevis-lmwg committed:
Merge remote-tracking branch 'escomp/master' into ne3np4_to_defaults_and_makefile

slevis-lmwg committed:
Replace comment explaining `allow_scm` with the correct one

slevis-lmwg committed:
Improve `allow_scm` comment further

slevis-lmwg committed:
Add comment explaining the read_noscm flag

slevis-lmwg committed:
Merge remote-tracking branch 'escomp/cesm3_0_beta04_changes' into ne3np4_to_defaults_and_makefile

slevis-lmwg committed:
Merge pull request #2835 from slevis-lmwg/ne3np4_to_defaults_and_makefile

ne3np4 to namelist_defaults_ctsm.xml and Makefile for PTS mode and add ability for warm starts in PTS mode

ekluzek committed:
Merge remote-tracking branch 'escomp/cesm3_0_beta04_changes' into fire_emis

ekluzek committed:
Change to full ESMF library

ekluzek committed:
Add tests for each of the Li fire types

ekluzek committed:
Add LAI stream to compile list

ekluzek committed:
Add check_dim_size

ekluzek committed:
Add FireClean method

ekluzek committed:
Make pointers set to NULL, and add BaseFireClean method

ekluzek committed:
Add shr_expr and fire_emis to list

ekluzek committed:
Add more setup and teardown, and also add several more tests, and include tests that test that fire emission on with nofire or SP fail, which right now they don't as expected

ekluzek committed:
Initialize pointer to NULL, add clean method, add some returns after endrun calls for unit testing

ekluzek committed:
Add a note about the call to the base BaseFireClean method failing at compile

ekluzek committed:
Get one of the failing tests working by implementing an init method that checks that fire-emission is off for fire_method==nofire

ekluzek committed:
Get the last test working by aborting if fire-emissions are on in SP mode

ekluzek committed:
Add CNFireSetFireMethod for unit testing of the CNFireFactory, removed need for reading in one namelist in the unit tester

ekluzek committed:
Initialize fire_method to UNSET, so it will fail if it isn't set by some means, and add two tests around for bad fire_method setting, one when NOT set, and another for a bad value that's set

ekluzek committed:
The unit testing actually doesn't need to call the fire_emis namelist read, as we can directly set fire-emission variables, which we SHOULD NOT BE ABLE TO DO! But, we'll take advantage of it for now

ekluzek committed:
Add FATES Fire factory files to list

ekluzek committed:
Add unit-tester for FATESFireFactory, two tests pass, two fail because they aren't implemented yet

ekluzek committed:
Add unit-tester for FATESFireFactory, two tests pass, two fail because they aren't implemented yet

ekluzek committed:
Get another test working, by checking that fire-emissions are off for FATES no_fire

ekluzek committed:
Merge tag 'ctsm5.3.030' into fire_emis

Fix FATES branch runs

ekluzek committed:
Don't allocate cnfire if NOT a BGC case

ekluzek committed:
Don't set stream types to NULL as they aren't pointers

ekluzek committed:
Make sure fire-emissions is off for FatesSp and FatesSt3 modes, which gets all of the tests working

ekluzek committed:
Add a fates_use_ed_st3 test and make sure it works

ekluzek committed:
Get new namelist tests to work

ekluzek committed:
Merge tag 'ctsm5.3.030' into cesm3_0_beta04_changes

Fix FATES branch runs

ekluzek committed:
Add failing test that both fates_sp and fates_st3 can be active at the same time

ekluzek committed:
Merge remote-tracking branch 'escomp/cesm3_0_beta04_changes' into fire_emis

ekluzek committed:
Change query of use_fates_sp to use nl_flags

This helps with #2643

ekluzek committed:
Another update to use nl_flags query for use_fates_sp

This also helps with #2643

[ Description of the changes in this commit. It should be enough
information for someone not following this development to understand.
Lines should be wrapped at about 72 characters. ]

ekluzek committed:
Don't allow both use_fates_sp and use_fates_ed_st3 to be on at the same time, die with an error

ekluzek committed:
Add checking around if ESMF is already initialized or not, and add note to NOT call ESMF_Finalize, see #3015 and #3017

ekluzek committed:
Refactor so FireInit doesn't include NLFilename or namelist reads, and FireReadNML is called seperate from FireInit

ekluzek committed:
Add some documentation about how some tests are standalone, and one is done as a loop. Also add a li2024crujra test, this is in response to code review

ekluzek committed:
Remove namelist file as no longer needed

ekluzek committed:
Improve the error message as noticed in the review

ekluzek committed:
Successful ignitions doesn't need to be specially handled

ekluzek committed:
Add a check to die when fire_emission and fates is on for now, the unit tests fail this way

ekluzek committed:
Get unit test working again

Get the FATESFireFactory tests working again now
that it aborts with fire-emission on.

ekluzek committed:
Fix typo

ekluzek committed:
Correct fire_emis option in CLM_BLD_NML_OPTS

Correct the build-namelist option for --no-fire_emis.
And also add more comments around this.

samsrabin commented:
@ekluzek What's the status of this? Is any of it ready to come in / can any of it be split out into different PRs?

ekluzek committed:
Fix fire initialization for full cases

Fire initialization needed some tweaks due to the refactoring
worked out in the unit testing.

ekluzek commented:
@samsrabin this is my next tag to bring in. We created this branch a long time ago because we thought that the PR's on it would flood b4b-dev too much. But, it turned out the work on it was spread out long enough that that wouldn't have been an issue. At least for this second go around. With b4b-dev we have a clock that ensures changes come in regularly, but we didn't have a process to that here. So it didn't happen. I think that means that in the future we should either slide these things as individual tags to master, or bring them to b4b-dev. Or think about the process that should be in place for a temporary branch like this....

It was brought into this branch as separate PR's similarly to b4b-dev and the first go around with this work. So you can view the contributions singly with those PR's already.

With your ask -- I thought about rebasing the PR's that went into this onto b4b-dev. But at this point I think that would only serve to create more work and slow the work down. So I won't do that now, but this can serve as something to think about as lessons for the future. I'll add this to the Thursday discussion.

ekluzek committed:
Put nofireemis first in the include file so that nofireemis is turned on, fixing #3073

ekluzek committed:
Merge tag 'ctsm5.3.040' into cesm3_0_beta04_changes

Update ctsm_pylib to 3.13.2

 Conflicts:
	cime_config/testdefs/testlist_clm.xml

ekluzek committed:
Merge pull request #3074 from ekluzek/update_cleanup_to_ctsm5340

Update cleanup branch to ctsm5340

ekluzek committed:
Merge remote-tracking branch 'escomp/cesm3_0_beta04_changes' into fire_emis

ekluzek committed:
Fix more Sp tests to have nofireemis

ekluzek committed:
Explicit check that use_cn is false so that drydepnomegan tests will work

ekluzek committed:
Add nofireemis to two more ClmSp tests

ekluzek committed:
Fix a few more tests that need to use nofireemis

ekluzek committed:
Merge tag 'ctsm5.3.043' into fire_emis

Merge b4bdev 20250509

Updates to the b4b-dev branch since its last merge to master (PRs #3091 #3092), as shown by git log:

- Merge pull request Update docs infrastructure #2809 from samsrabin/update-docs-builder-2
  Update docs infrastructure
- Merge pull request Update externals to cesm3_0_alpha06c #3106 from ekluzek/update_to_alpha06c
  Update externals to cesm3_0_alpha06c
- Merge pull request User control over snow thermal conductivity scheme over glaciers #3072 from wwieder/JordanGlacier
  User control over snow thermal conductivity scheme over glaciers

 Conflicts:
	src/biogeochem/CNFireNoFireMod.F90

ekluzek committed:
Allocate cnfire_method for fates as nofire, because it's passed down to subroutines that FATES needs

ekluzek committed:
Merge tag 'ctsm5.3.053' into fire_emis

Fix and improve anomaly forcings for ISSP cases

 Conflicts:
	bld/unit_testers/build-namelist_test.pl
	src/biogeochem/SatellitePhenologyMod.F90

ekluzek committed:
Fix more tests with sp that need nofireemis

ekluzek committed:
Another nofireemis change

ekluzek committed:
Merge tag 'ctsm5.3.057' into cesm3_0_beta04_changes

Fix PEM test for FATES testmod that builds an on-the-fly parameter file

ekluzek commented:
I updated to ctsm5.3.057 and testing for aux_clm is passing as expected on Derecho and Izumi.

ekluzek committed:
Merge branch 'cesm3_0_beta04_changes' into fire_emis

ekluzek committed:
Merge pull request #2844 from ekluzek/fire_emis

branch with 5.3.05Y: Make sure FORTRAN code properly aborts when fire-emission is asked for and it can't be provided

slevis-lmwg commented on src/biogeochem/CNFireBaseMod.F90:
```diff
@@ -392,9 +416,11 @@ subroutine FireReadNML( this, NLFilename )
              end if
           else
              call endrun(msg="ERROR could NOT find "//nmlname//"namelist"//errmsg(sourcefile, __LINE__))
+             return
```
Do "return" statements serve a purpose after an endrun?

slevis-lmwg reviewed (approved):
I'm reviewing the part of this that comes in with #2844 since the other PRs were mine.

The title of #2844 predisposed me to expect a short PR with few code changes. This turns out to be a major (in my perception) refactor, plus an addition of unit tests. Thank you for all this @ekluzek!

I included a few questions. Nothing stood out as problematic to me.

ekluzek commented on src/unit_test_stubs/main/ncdio_pio_fake.F90.in:
```diff
@@ -340,6 +341,25 @@ contains
+    ! !LOCAL VARIABLES:
+    !-----------------------------------------------------------------------
+
+    ! Does nothing assumes the dim size is as expected
```
This is a stub used for the unit testing. The stub module already existed, I just needed to add a new interface that wasn't previously used.

What we are doing in the PFUnit testing is allowing ncdio_pio subroutine calls to be made -- but they do nothing. This removes the problems in unit-testing that filesystem read/writes add. Which is good to do to have very fast solid and robust testing.

ekluzek committed:
Merge tag 'ctsm5.3.058' into cesm3_0_beta04_changes

Fix clm6 ISSP compset aliases

ekluzek committed:
Get the parse cs.status script to work in a CESM checkout

ekluzek committed:
Change directory name of mizuroute to match what is in CESM

ekluzek commented:
I ran some extra tests to make sure the issues were resolved. These PASS:

PEA.1x1_smallvilleIA.IHistClm50BgcCropQianRs.derecho_gnu.clm-smallville_dynurban_monthly			
PEA_D.1x1_smallvilleIA.IHistClm50BgcCropQianRs.derecho_gnu.clm-smallville_dynurban_monthly			
SEQ_D_PS.f09_f09_mt232.I1850Clm50Sp.derecho_intel.clm-default--clm-nofireemis			
SEQ_PS.f09_f09_mt232.I1850Clm50Sp.derecho_intel.clm-default--clm-nofireemis			
SMS_D_Ln9.f19_f19_mg17.FWma2000climo.derecho_intel.cam-outfrq9s_waccm_ma_mam4			
SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCHIST.derecho_intel.cam-outfrq9s			
SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCnudged.derecho_intel.cam-outfrq9s			
SMS_Ld12_Mmpi-serial.1x1_vancouverCAN.I1PtClm60SpRs.derecho_gnu.clm-output_sp_highfreq			
SMS_Ln9.ne3pg3_ne3pg3_mg37.I2000Clm50Sp.derecho_gnu.clm-clm50cam6LndTuningMode--clm-nofireemis			
SMS_Ly1_Mmpi-serial.1x1_brazil.IHistClm60BgcQianRs.derecho_intel.clm-output_bgc_highfreq			
SSPMATRIXCN_Ly5_Mmpi-serial.1x1_numaIA.I2000Clm60BgcCropQianRs.derecho_intel.clm-ciso_monthly

These fail and need to be figured out:

ERS_Ld3.f19_f19_mg17.FXHIST.derecho_intel.cam-waccmx_weimer	(SETUP)		
SMS_D_Ln9.f19_f19_mg17.FXHIST.derecho_intel.cam-outfrq9s_amie	(SETUP)		
SMS_D_Ln9_P1280x1.ne0ARCTICne30x4_ne0ARCTICne30x4_mt12.FHIST.derecho_intel.cam-outfrq9s	(NLCOMP RUN)

ekluzek committed:
Change name for mizuRoute so it's all lower case the same as it is in CESM

ekluzek commented:
It looks like the problem with the f19 tests are the unusual RUN_STARTDATE of: 2005-12-31 and 2003-10-28. It fails with the problem we have been seeing where use_init_interp isn't coordinated correctly with the IC files.

The ARCTIC grid fails in an ESMF regrid and I think it's likely due to too few processors. The default for CAM is 91 nodes, which is 11k tasks, so 10X larger than the tasks asked for. So I think it will likely work with more processors.

ekluzek commented:
Since, I had some variety of tests failing, I sent the ctsm_sci test list and these tests failed as they need the --clm-nofireemis added as an extra testmod to them.

SMS_Ld5.f09_g17.IHistClm50SpCru.derecho_intel.clm-default       (SETUP)
SMS_Ld5.f19_g17.IHistClm50SpCru.derecho_intel.clm-default       (SETUP)
SMS_Lm12.f09_f09_mg17.I1850Clm60Sp.derecho_intel.clm-ExcessIceStartup_output_sp_exice   (SETUP)
SMS_Lm12.f09_t232.I1850Clm60SpCrujra.derecho_intel.clm-ExcessIceStartup_output_sp_exice (SETUP)

ekluzek commented:
In the meeting this morning, I was encouraged to make the tag as is. And file issues for anything outstanding. I have created baselines that just need to be renamed. And I'll start the ChangeLog.

ekluzek committed:
Update change files

ekluzek committed:
Fix a couple ctsm_sci tests that were failing as they need nofiremis testmod added

ekluzek committed:
Adjust some tests as per #2863 adding a few more to aux_cime_baselines, adding SEQ and PEA tests, make sure the changes are for Clm60 tests

ekluzek committed:
Add an izumi_nag mpi-serial non-DEBUG test for aux_clm and prealpha

ekluzek committed:
Actually we don't want any non-DEBUG izumi_nag tests

ekluzek committed:
Add some tests to aux_cime_baselines to cover more build types

ekluzek committed:
Add a threaded derecho_gnu test to prebeta and aux_cime_baselines

ekluzek committed:
Final update of change files
